README.md
# sfecol - Colors of the San Francisco Estuary
***
The San Francisco Estuary is a complex watershed and home to many different species of wildlife. This `R` package uses images of the landscape and wildlife to create color palettes. Colors are extracted from images using https://colordesigner.io/color-palette-from-image. 
This package closely imitates the excellent work of the [`Manu`](https://g-thomson.github.io/Manu) and [`wesanderson`](https://github.com/karthik/wesanderson) packages.
 
## Installation
***
This package is hosted on Github and can install it using the `devtools` package:
``` r
# install.packages("devtools")
devtools::install_github("MalteWillmes/sfecol")
```

## Usage
***
The color palettes are stored as a list named `sfe_palettes`. Thus you can see a list of the available palettes like so:
```r
library(sfecol)
names(sfe_palettes)
[1] "delsme" "chinook" "chinookcarc" "southbay" "baslu" "elpastel"   
```

A helper function `sfe_pal()` returns the desited color palette as a vector: For example:
```r
sfe_pal("delsme")
[1] "#304247" "#D1D0B3" "#828D6D" "#8DBA71" "#92A2AF" "#447484"
```
There is also a helper function called `print_pal()` which displays the palette in the graphics window.
```r
print_pal(sfe_pal("delsme"))
```
<img src="images/delsme_palette.png" widht=30%/>

Since the `get_pal()` function returns the color palettes as a character vector they can easily be used in the graphics package of your choice. Here are examples in both base `R` and [`ggplot2`](https://ggplot2.tidyverse.org/).
```r
# Base R implementation
plot(mtcars$disp, mtcars$hp, col = sfe_pal("delsme")[factor(mtcars$carb)], pch = 19, cex = 3 )

# ggplot2 implementation
library(ggplot2)
ggplot(mtcars, aes(disp, hp, color = factor(carb))) +
   geom_point(size = 3) +
   scale_color_manual(values = sfe_pal("delsme"))
```

## Color palettes
***
The data for the graphs shown below comes from the Palmerpenguins dataset. This dataset contains various measurments of 344 penguins from 3 different species of penguins, collected from 3 islands in the Palmer Archipelago, Antarctica. 

Horst AM, Hill AP, Gorman KB (2020). palmerpenguins: Palmer Archipelago (Antarctica) penguin data. R package version 0.1.0. https://allisonhorst.github.io/palmerpenguins/. doi: 10.5281/zenodo.3960218.

### delsme - *Hypomesus transpacificus* - Delta Smelt
```r
c("#304247", "#D1D0B3", "#828D6D", "#8DBA71", "#92A2AF","#447484")
```
<img src="images/delsme.png" width=100%/>
Image: The Delta Smelt has a typical cucumber smell to it. Photo by Peterson, B. Moose, U.S. Fish and Wildlife Service.

### chinook - *Oncorhynchus tshawytscha* - Chinook Salmon
```r
c("#899FBF", "#243158", "#543D4A", "#D8D9E4", "#715E4D", "#C0D5F0")
```
<img src="images/chinook.png" width=100%/>
Image: A winter run Chinook Salmon in Battle Creek, CalTrout (https://caltrout.org/). Photo by Jane Work.

### chinookcarc - *Oncorhynchus tshawytscha* - Chinook Salmon carcass on Putah Creek
```r
c("#CDC9C0", "#2B2F2A", "#837C32", "#727877", "#634933", "#A89C6B")
```
<img src="images/chinookcarc.png" width=100%/>
Image: A Chinook Salmon carcass on Putah Creek. Photo by Ken Davis.

### southbay  - Soutbay Salt Ponds
```r
c("#727467", "#97CFCB", "#D5A54C", "#1F2F3D", "#801F23", "#633F47")
```
<img src="images/southbay.png" width=100%/>
Image: Bird's-eye view of the multi-colored salt ponds, Photo by Doc Searls from Santa Barbara, USA. 


### petaluma  - Petaluma River
```r
c("#DFDCD0" "#37322A" "#94908B" "#7A6863" "#857C6C" "#6C7C7E")
```
<img src="images/petaluma.png" width=100%/>
Image: Survey boat on the Petaluma River. Photo by Malte Willmes.

### napa  - Napa River
```r
c("#D6C8B9" "#181618" "#8F959A" "#73615D" "#A64F40" "#4F575B")
```
<img src="images/napa.png" width=100%/>
Image: Early morning on the Napa River. Photo by Malte Willmes.

### baslu - *Ariolimax columbianus* - Banana slug
```r
c("#6F7527", "#D0C56E", "#122006", "#AEB953","#3F2D0F","#9B9899")
```
<img src="images/baslu.png" width=100%/>
Image: A banana slug. Photo by Ben Stanfield.

## Color palettes not based on images
***
Here are additional color palettes that are not based on images but have proven to be useful to show SFE related data

### elpastel - pastel color schema for discrete data
```r
sfe_pal("elpastel")
[1] "#4E79A7" "#F28E2B" "#E15759" "#76B7B2" "#59A14F" "#EDC948" "#B07AA1" "#FF9DA7" "#9C755F" "#BAB0AC"
print_pal(sfe_pal("elpastel"))
```
<img src="images/elpastel_palette.png" width=50%/>

### delsmelife - color schema from Hobbs et al., 2019
```r
sfe_pal("delsmelife")
[1] "#8FBC8F" "#A020F0" "#FEF0D9" "#FDCC8A" "#FC8D59" "#D7301F"
print_pal(sfe_pal("delsmelife"))
```
<img src="images/delsmelife_palette.png" width=50%/>

## Continuous color palettes
***
The color palettes in this package are designed for discrete variables. However if your data is continuous and needs to be plotted as such (e.g. heatmaps) you can use the `colorRampPalette()` already part of your `R` installation to create a color gradient.

For example, if you would like to create a color gradient between the 2nd and 3rd colors from the `delsme` palette you could do the following.
```r
# Select 2nd and 3rd colors
selected_colors <- sfe_pal("delsme")[c(2,3)]
# Create a gradient of 100 colours between the selected colours
colorRampPalette(selected_colors)(100)
```


NEWS.md
# sfecol 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/sfecol.R
#' @title Color palettes based on the San Francisco Estuary.
#' @description This package provides color palettes based on images from the San Francisco Estuary. Use \code{\link{sfe_pal}} extract palettes.
#' Package based on the excellent manu package (https://g-thomson.github.io/Manu/) by Dr Tara McAllister and Karthik Ram
#' Complete list of palettes
#'
#'
#' @export
sfe_palettes <- list(
  delsme= c("#304247", "#D1D0B3", "#828D6D", "#8DBA71", "#92A2AF","#447484"),
  chinook= c("#899FBF", "#243158", "#543D4A", "#D8D9E4", "#715E4D", "#C0D5F0"),
  chinookcarc= c("#CDC9C0", "#2B2F2A", "#837C32", "#727877", "#634933", "#A89C6B"),
  southbay= c("#727467", "#97CFCB", "#D5A54C", "#1F2F3D", "#801F23", "#633F47"),
  baslu= c("#6F7527", "#D0C56E", "#122006", "#AEB953","#3F2D0F","#9B9899"),
  petaluma= c("#DFDCD0","#37322A","#94908B","#7A6863","#857C6C","#6C7C7E"),
  napa=c("#D6C8B9","#181618","#8F959A","#73615D","#A64F40","#4F575B"),
  elpastel= c('#4E79A7', '#F28E2B','#E15759','#76B7B2','#59A14F','#EDC948','#B07AA1','#FF9DA7','#9C755F','#BAB0AC'),
  delsmelife =c("#8FBC8F","#A020F0", "#FEF0D9", "#FDCC8A" ,"#FC8D59" ,"#D7301F")
)



#' Select SFE color palette
#'
#' This function returns color palettes derived from photos of the San Francisco Estuary.
#'
#' @param name Name of desired palette. Choices are: \code{delsme}, \code{chinook},  \code{chinookcarc}, \code{southbay},
#' \code{baslu},\code{petaluma}, \code{napa},\code{elpastel}, \code{delsmelife}
#'
#' @return A vector of colors.
#' @export
#' @examples
#' sfe_pal("delsme")
#' sfe_pal("chinook")
#' sfe_pal("chinookcarc")
#' sfe_pal("southbay")
#' sfe_pal("petaluma")
#' sfe_pal("napa")
#' sfe_pal("baslu")
#' sfe_pal("elpastel")
#' sfe_pal("delsmelife")
#'
sfe_pal <- function(name) {

  pal <- sfe_palettes[[name]]
  if (is.null(pal))
    stop("Palette not found.")

  pal
}

#' Plots a color palette
#'
#' @param x a vector of colors
#' @export
#' @importFrom graphics par image
print_pal <- function(x, ...) {
  n <- length(x)
  user_par = par(no.readonly = T)
  par(mfrow = c(1,1), mar = c(rep(0.5,4)), oma =c(rep(0,4)))
  on.exit(par(user_par))

  image(1:n, 1, as.matrix(1:n), col = x,
        ylab = "", xaxt = "n", yaxt = "n", bty = "n")
}
